### PR TITLE
Improve performance of editorplaceholder plugin

### LINKED ISSUE #5184: Performance degradation for large documents with "editorplaceholder" plugin enabled
## Type of report

Bug

## Provide detailed reproduction steps (if any)

Plugin "editorplaceholder" brings noticeable lagging while typing text in large documents.

I prepared a minimal configuration sandbox https://codepen.io/ckeditorbug/pen/oNEjBWL. Used the latest 4.18.0 "full" build CKEditor version.
I've used 3mb content to reproduce the issue. Performance degradation is noticeable with a smaller content size but 3mb was right enough for me to clearly see the difference.

1. Enable "editorplaceholder" plugin through CKEditor configuration. I believe it's enabled by default in "full" distribution we use.
2. Paste large content to the editor.
3. Type any sentence in the editor.
4. Observe very laggy typing experience.

I did some performance measurements:
plugin is enabled
Every "keyup" event with enabled "editorplaceholder" takes around 140-250ms (depends on a PC configuration).

After disabling the "editorplaceholder" plugin I'm getting much better typing experience. Which is proved by performance chart:
plugin is disabled

Some observations. "editorplaceholder" registers the on "change" event listener. It takes long time to execute the code of the event handler "togglePlaceholder" https://github.com/ckeditor/ckeditor4/blob/master/plugins/editorplaceholder/plugin.js#L86.

### Expected result

"editorplaceholder" plugin should not affect typing experience that much.

### Actual result

Laggy typing experience. Long response on keyboard events.

## Other details

* Browser: Latest Chrome, Firefox, Edge.
* OS: Windows 11.
* CKEditor version: 4.18.0
* Installed CKEditor plugins: default configuration, used full build version.

### PR DESCRIPTION
**Pull request by jacekbogdanski**
## What is the purpose of this pull request?

Bug fix

## Does your PR contain necessary tests?

All patches that change the editor code must include tests. You can always read more
on [PR testing](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_contributing_code.html#tests),
[how to set the testing environment](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html) and
[how to create tests](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html#creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [x] Unit tests
- [ ] Manual tests

## Did you follow the CKEditor 4 code style guide?

Your code should follow the guidelines from the [CKEditor 4 code style guide](https://github.com/ckeditor/ckeditor4/blob/major/dev/docs/codestyle.md) which helps keep the entire codebase consistent.

- [x] PR is consistent with the code style guide

## What is the proposed changelog entry for this pull request?

```
Fixed:
* [#5184](https://github.com/ckeditor/ckeditor4/issues/5184): Fixed: [Editor Placeholder](https://ckeditor.com/cke4/addon/wysiwygarea) plugin degredates typing performance.

API changes:
* [#5184](https://github.com/ckeditor/ckeditor4/issues/5184): Added [`config.editorplaceholder_delay`](https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_config.html#cfg-editorplaceholder_delay) configuration option allowing to delay placeholder before it is toggled when changing editor content.
* [#5184](https://github.com/ckeditor/ckeditor4/issues/5184): Added [`CKEDITOR.tools.debounce()`](https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_tools.html#method-debounce) function allowing to postpone passed function execution until the given milliseconds have elapsed since the last time it was invoked
```

## What changes did you make?

Added simple debounce logic allowing to mitigate the issue with toggling editor placeholder on change event too often. I didn't add any manual tests as I don't really see any deterministic way to manually test that change, since the typing performance may be connected to a specific machine. Most likely, unit tests should be sufficient to test these changes.

## Which issues does your PR resolve?

Closes #5184

### CODE REVIEW

File: plugins/editorplaceholder/plugin.js [hunk 1]
```diff
@@ -21,13 +21,12 @@
...
 
-			bindPlaceholderEvents( editor, [
-				'contentDom',
-				'focus',
-				'blur',
-				'change'
-			] );
+			bindPlaceholderEvent( editor, 'contentDom' );
+			bindPlaceholderEvent( editor, 'focus' );
+			bindPlaceholderEvent( editor, 'blur' );
```
Comment (jacekbogdanski) [inline plugins/editorplaceholder/plugin.js:26]: These events are not debounced on purpose, as debounce logic on focus/blur creates a visible placeholder delay. In probably 99% of cases we could simply skip toggling editor placeholder on the `change` event as, during normal editor usage, the editor placeholder is not visible until you blur the editor. However, a `change` event is required for API manipulation like `editor.insertText`.

**Comment by Comandeer** [comment]
Rebased onto latest `master`.

**Review by Comandeer — CHANGES_REQUESTED** [review]
WDYT about adding some manual test, e.g. an editor with a big image inside? It should allow to verify if the fix really fix the performance issue.

Additionally, we already have [`CKEDITOR.tools.buffers.throttle`](https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_tools_buffers_throttle.html). I'm wondering if it won't be more consistent if the newly added API was added as `CKEDITOR.tools.buffers.debounce` and has a similar shape to other `CKEDITOR.tools.buffers` members, instead of being a completely separate being.

File: plugins/editorplaceholder/plugin.js [hunk 2]
```diff
@@ -58,10 +57,14 @@
...
-	function bindPlaceholderEvents( editor, events ) {
-		CKEDITOR.tools.array.forEach( events, function( event ) {
-			editor.on( event, togglePlaceholder, null, { editor: editor } );
-		} );
+	function bindPlaceholderEvent( editor, eventName, delay ) {
+		var toggleFn = togglePlaceholder;
+
+		if ( delay ) {
+			toggleFn = CKEDITOR.tools.debounce( togglePlaceholder, delay );
+		}
```
[the remark this answers is not in the document]
Comment (jacekbogdanski) [reply plugins/editorplaceholder/plugin.js:65]: What is the advantage of such change? It sounds like a nitpick for me (which could even decrease readability, but that's subjective).

**Comment by jacekbogdanski** [comment]
> WDYT about adding some manual test, e.g. an editor with a big image inside? It should allow to verify if the fix really fix the performance issue.

I'm a bit skeptical of that proposition because it still won't let us test these changes in a deterministic manner. Manual tests are run basically only in 2 cases:
* PR review
* Testing phase
As for PR review, I agree that adding such test could improve the overall reviewing experience, as you can just check out the old version, keep the manual test, and confirm the results. However, in the case of the testing phase, we may get many false-positive results e.g because manual tests will rely on personal tester experience (the high-resolution images will slow down the editor anyway).

I've checked how it would work with a solid base64 2.5MB image and I can admit that there is a difference in typing performance, but not that significant to differentiate it without direct comparison to code with no patch.

**Comment by jacekbogdanski** [comment]
> Additionally, we already have [CKEDITOR.tools.buffers.throttle](https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_tools_buffers_throttle.html). I'm wondering if it won't be more consistent if the newly added API was added as CKEDITOR.tools.buffers.debounce and has a similar shape to other CKEDITOR.tools.buffers members, instead of being a completely separate being.

While I agree that debounce somehow fits those classes, I would like to avoid such dedication to mimic the similar API, as currently the debounce function is much simpler than what we have in `CKEDITOR.tools.buffers`.

**Comment by Comandeer** [comment]
> However, in the case of the testing phase, we may get many false-positive results e.g because manual tests will rely on personal tester experience (the high-resolution images will slow down the editor anyway).

WDYT about adding a manual test with two editors – one with the `editorplaceholder` plugin and one without it? This way the developer would be able to compare the performance between them and check if the plugin degrades it. That would allow us to empirically validate if the fix really works.

**Comment by Comandeer** [comment]
Rebased onto latest `master`.

**Review by Comandeer — APPROVED** [review]
LGTM!

**Pull request merged**